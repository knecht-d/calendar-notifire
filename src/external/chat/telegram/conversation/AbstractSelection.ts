import Telegraf, { Context } from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";

export abstract class AbstractSelection<T> {
    constructor(
        protected editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
        protected messageId: number,
    ) {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    static setActions(_bot: Telegraf<TelegrafContext>, _selectionHandler: (ctx: Context, value: string) => void) {}
    abstract requestInput(buildingObject: T): Promise<(selection: string) => T>;
}
