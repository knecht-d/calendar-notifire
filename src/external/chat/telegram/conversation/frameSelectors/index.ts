export { DaySelection } from "./DaySelection";
export { HourSelection } from "./HourSelection";
export { MinuteSelection } from "./MinuteSelection";
export { MonthSelection } from "./MonthSelection";
