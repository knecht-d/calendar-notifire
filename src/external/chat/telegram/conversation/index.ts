export * from "./FrameConfigBuilder";
export * from "./TriggerConfigBuilder";
