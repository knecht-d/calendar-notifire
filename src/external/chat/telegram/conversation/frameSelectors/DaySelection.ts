import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericFrame } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";
import { splitInChunks } from "../utils";

const prefix = "FRAME_DAY";
const values = [...[...Array(10).keys()].map(x => `+${x + 1}`)];
const skip = "SKIP";

export class DaySelection extends AbstractSelection<IGenericFrame> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}|${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
        bot.action(`${prefix}|${skip}`, ctx => selectionHandler(ctx, `${skip}`));
    }

    async requestInput(buildingObject: IGenericFrame): Promise<(selection: string) => IGenericFrame> {
        await this.editMessage(
            "Bis zu welchem Tag sollen die Termine betrachtet werden?",
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    [
                        ...splitInChunks(
                            values.map(value => Markup.callbackButton(value, `${prefix}|${value}`)),
                            5,
                        ),
                        [Markup.callbackButton("Aktueller Tag", `${prefix}|${skip}`)],
                    ],
                    { columns: 5 },
                ),
            ),
        );
        return (selection: string) => {
            if (selection !== skip) {
                buildingObject.day = {
                    fixed: false,
                    value: Number.parseInt(selection),
                };
            }
            return buildingObject;
        };
    }
}
