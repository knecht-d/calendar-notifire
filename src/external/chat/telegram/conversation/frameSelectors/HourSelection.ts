import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericFrame } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const prefix = "FRAME_HOUR";
const values = [...[...Array(24).keys()].map(x => `${x}`), ...[...Array(12).keys()].map(x => `+${x + 1}`)];
const skip = "SKIP";

export class HourSelection extends AbstractSelection<IGenericFrame> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}|${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
        bot.action(`${prefix}|${skip}`, ctx => selectionHandler(ctx, `${skip}`));
    }

    async requestInput(buildingObject: IGenericFrame): Promise<(selection: string) => IGenericFrame> {
        await this.editMessage(
            `
Bis zu welcher Stunde sollen die Termine betrachtet werden?

+/- und "Aktuelle Stunde" bedeuten, dass der Wert relativ zum Zeitpunkt der Erinnerung betrachtet wird.
z.B. "-5" Stunden um 17:00 , dass Termine ab 12:00 betrachtet werden.`.trim(),
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    [
                        ...values.map(option => Markup.callbackButton(option, `${prefix}|${option}`)),
                        Markup.callbackButton("Aktuelle Stunde", `${prefix}|${skip}`),
                    ],
                    { columns: 6 },
                ),
            ),
        );
        return (selection: string) => {
            if (selection !== skip) {
                buildingObject.hour = {
                    fixed: !selection.startsWith("+"),
                    value: Number.parseInt(selection),
                };
            }
            return buildingObject;
        };
    }
}
