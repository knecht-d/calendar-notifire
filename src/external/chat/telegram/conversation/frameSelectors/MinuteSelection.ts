import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericFrame } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const prefix = "FRAME_MINUTE";
const values = ["0", "15", "30", "45", "59", "+15", "+30", "+45"];
const skip = "SKIP";

export class MinuteSelection extends AbstractSelection<IGenericFrame> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}|${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
        bot.action(`${prefix}|${skip}`, ctx => selectionHandler(ctx, `${skip}`));
    }

    async requestInput(buildingObject: IGenericFrame): Promise<(selection: string) => IGenericFrame> {
        await this.editMessage(
            `
Bis zu welcher Minute sollen die Termine betrachtet werden?

+/- und "Aktuelle Minute" bedeuten, dass der Wert relativ zum Zeitpunkt der Erinnerung betrachtet wird.
z.B. "+30" Minuten um 12:00 , dass Termine ab 12:30 betrachtet werden.`.trim(),
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    [
                        [
                            ...values
                                .filter(option => !option.startsWith("+"))
                                .map(option => Markup.callbackButton(option, `${prefix}|${option}`)),
                        ],

                        [
                            ...values
                                .filter(option => option.startsWith("+"))
                                .map(option => Markup.callbackButton(option, `${prefix}|${option}`)),
                        ],
                        [Markup.callbackButton("Aktuelle Minute", `${prefix}|${skip}`)],
                    ],
                    { columns: 4 },
                ),
            ),
        );
        return (selection: string) => {
            if (selection !== skip) {
                buildingObject.minute = {
                    fixed: !selection.startsWith("+"),
                    value: Number.parseInt(selection),
                };
            }
            return buildingObject;
        };
    }
}
