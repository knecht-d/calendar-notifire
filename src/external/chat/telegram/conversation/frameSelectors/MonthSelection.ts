import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericFrame } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const prefix = "FRAME_MONTH";
const values = ["+1", "+2", "+3"];
const skip = "SKIP";

export class MonthSelection extends AbstractSelection<IGenericFrame> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}|${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
        bot.action(`${prefix}|${skip}`, ctx => selectionHandler(ctx, `${skip}`));
    }

    async requestInput(buildingObject: IGenericFrame): Promise<(selection: string) => IGenericFrame> {
        await this.editMessage(
            "Bis zu welchem Monat sollen die Termine betrachtet werden?",
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    [
                        [...values.map(option => Markup.callbackButton(option, `${prefix}|${option}`))],
                        [Markup.callbackButton("Aktueller Monat", `${prefix}|${skip}`)],
                    ],
                    { columns: 2 },
                ),
            ),
        );
        return (selection: string) => {
            if (selection !== skip) {
                buildingObject.month = {
                    fixed: !selection.startsWith("+"),
                    value: Number.parseInt(selection),
                };
            }
            return buildingObject;
        };
    }
}
