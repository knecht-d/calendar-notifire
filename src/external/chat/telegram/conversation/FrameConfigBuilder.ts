import Telegraf from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericFrame, IGenericRecurrence } from "../../../../gateways";
import { ILogger } from "../../../logging";
import { AbstractSelection } from "./AbstractSelection";
import { DaySelection, HourSelection, MinuteSelection, MonthSelection } from "./frameSelectors";
import { handleButtonPress } from "./utils";

export class FrameConfigBuilder {
    private static ALL_BUILDER: { [messageId: number]: FrameConfigBuilder } = {};
    private static LOGGER: ILogger;
    config: IGenericFrame;
    currentResolver: (selection: string) => void;

    constructor(
        private messageId: number,
        private editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
    ) {
        this.config = {};
        this.currentResolver = () => {
            return;
        };
        FrameConfigBuilder.ALL_BUILDER[messageId] = this;
    }

    static init(bot: Telegraf<TelegrafContext>, logger: ILogger) {
        this.LOGGER = logger;
        [MonthSelection, DaySelection, MinuteSelection, HourSelection].forEach(selection => {
            selection.setActions(bot, this.handleButtonPress.bind(this));
        });
    }

    static handleButtonPress(ctx: TelegrafContext, selection: string) {
        handleButtonPress(this.ALL_BUILDER, this.LOGGER, ctx, selection);
    }

    async requestConfig(recurrence: IGenericRecurrence) {
        switch (recurrence.type) {
            case "m":
                await this.requestSelection(MonthSelection);
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                break;
            case "d":
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            case "h":
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            default:
                throw new Error("Unexpected Recurrence Type");
        }

        delete FrameConfigBuilder.ALL_BUILDER[this.messageId];
        return this.config;
    }

    getConfig() {
        return this.config;
    }

    private async requestSelection(
        Selection: new (
            editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
            messageId: number,
        ) => AbstractSelection<IGenericFrame>,
    ) {
        const selection = new Selection(this.editMessage, this.messageId);
        const selectionHandler = await selection.requestInput(this.config);
        return new Promise<void>(res => {
            this.currentResolver = selection => {
                this.config = selectionHandler(selection);
                res();
            };
        });
    }
}
