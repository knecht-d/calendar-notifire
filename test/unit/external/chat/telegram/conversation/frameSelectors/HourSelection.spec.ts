import { Extra, Markup } from "telegraf";
import { HourSelection } from "../../../../../../../src/external/chat/telegram/conversation/frameSelectors";

describe("HourSelection", () => {
    const fixedValues = [
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "11",
        "12",
        "13",
        "14",
        "15",
        "16",
        "17",
        "18",
        "19",
        "20",
        "21",
        "22",
        "23",
    ];
    const relativeValues = ["+1", "+2", "+3", "+4", "+5", "+6", "+7", "+8", "+9", "+10", "+11", "+12"];
    const values = [...fixedValues, ...relativeValues];
    const skip = "SKIP";
    const mockCtx = { mock: "ctx" };
    const MockBot = {
        action: jest.fn((_, handler) => {
            handler(mockCtx);
        }),
    };
    const mockEditMessage = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    describe("setActions", () => {
        it.each([...values, skip])("should set action for %s", value => {
            const mockSelectionHandler = jest.fn();
            HourSelection.setActions(MockBot as any, mockSelectionHandler);
            expect(MockBot.action).toHaveBeenCalledWith(`FRAME_HOUR|${value}`, expect.any(Function));
        });

        it.each([...values, skip])("should set the correct handler for %s", value => {
            const mockSelectionHandler = jest.fn();
            HourSelection.setActions(MockBot as any, mockSelectionHandler);
            expect(mockSelectionHandler).toHaveBeenCalledWith(mockCtx, value);
        });
    });

    describe("requestInput", () => {
        it("should send the question", async () => {
            const selection = new HourSelection(mockEditMessage, 42);
            await selection.requestInput({});
            expect(mockEditMessage).toHaveBeenCalledWith(
                expect.stringContaining("Bis zu welcher Stunde "),
                expect.any(Object),
            );
        });

        it("should have the correct keyboard", async () => {
            jest.spyOn(Extra, "markup").mockImplementation((): any => {
                return {
                    markup: (cb: () => any) => cb(),
                };
            });
            jest.spyOn(Markup, "callbackButton").mockImplementation(
                (desc: string, key: string): any => `${desc}(${key})`,
            );
            jest.spyOn(Markup, "inlineKeyboard");

            const selection = new HourSelection(mockEditMessage, 42);
            await selection.requestInput({});

            expect((Markup.inlineKeyboard as jest.Mock).mock.calls).toMatchSnapshot();
        });

        it.each(fixedValues)("should set the fixed hour on selection of %s", async value => {
            const selection = new HourSelection(mockEditMessage, 42);
            const handler = await selection.requestInput({});
            const result = handler(value);
            expect(result).toEqual({
                hour: {
                    fixed: true,
                    value: fixedValues.indexOf(value),
                },
            });
        });

        it.each(relativeValues)("should set the relative hour on selection of %s", async value => {
            const selection = new HourSelection(mockEditMessage, 42);
            const handler = await selection.requestInput({});
            const result = handler(value);
            expect(result).toEqual({
                hour: {
                    fixed: false,
                    value: relativeValues.indexOf(value) + 1,
                },
            });
        });

        it("should not set the hour on selection of SKIP", async () => {
            const selection = new HourSelection(mockEditMessage, 42);
            const handler = await selection.requestInput({});
            const result = handler(skip);
            expect(result).toEqual({});
        });
    });
});
