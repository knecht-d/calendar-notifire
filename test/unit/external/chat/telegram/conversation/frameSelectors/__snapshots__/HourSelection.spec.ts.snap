// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourSelection requestInput should have the correct keyboard 1`] = `
Array [
  Array [
    Array [
      "0(FRAME_HOUR|0)",
      "1(FRAME_HOUR|1)",
      "2(FRAME_HOUR|2)",
      "3(FRAME_HOUR|3)",
      "4(FRAME_HOUR|4)",
      "5(FRAME_HOUR|5)",
      "6(FRAME_HOUR|6)",
      "7(FRAME_HOUR|7)",
      "8(FRAME_HOUR|8)",
      "9(FRAME_HOUR|9)",
      "10(FRAME_HOUR|10)",
      "11(FRAME_HOUR|11)",
      "12(FRAME_HOUR|12)",
      "13(FRAME_HOUR|13)",
      "14(FRAME_HOUR|14)",
      "15(FRAME_HOUR|15)",
      "16(FRAME_HOUR|16)",
      "17(FRAME_HOUR|17)",
      "18(FRAME_HOUR|18)",
      "19(FRAME_HOUR|19)",
      "20(FRAME_HOUR|20)",
      "21(FRAME_HOUR|21)",
      "22(FRAME_HOUR|22)",
      "23(FRAME_HOUR|23)",
      "+1(FRAME_HOUR|+1)",
      "+2(FRAME_HOUR|+2)",
      "+3(FRAME_HOUR|+3)",
      "+4(FRAME_HOUR|+4)",
      "+5(FRAME_HOUR|+5)",
      "+6(FRAME_HOUR|+6)",
      "+7(FRAME_HOUR|+7)",
      "+8(FRAME_HOUR|+8)",
      "+9(FRAME_HOUR|+9)",
      "+10(FRAME_HOUR|+10)",
      "+11(FRAME_HOUR|+11)",
      "+12(FRAME_HOUR|+12)",
      "Aktuelle Stunde(FRAME_HOUR|SKIP)",
    ],
    Object {
      "columns": 6,
    },
  ],
]
`;
